### PR TITLE
BE: parameters in snippet

### PR DESCRIPTION
**Pull request by qnkhuat**
Add `template_tags` column to snippet and make substitution middleware able to substitute parameters in snippet.

### Background

The simplified idea of our parameter substitution works as follows:
- The `substitute-parameters` middleware receives a native query and triggers a driver method `substitute-native-parameters` - which does the actual substitution
The code of this driver method
```clojure
(sql.params.substitute/substitute 
  (params.parse/parse query)
  (params.values/query->params-map inner-query))
```

1. the `params.parse/parse` function takes a raw query (e.g: select * from {{table-name}}) and try to parse it, the result will be an array where each elements is either a raw string, or a Param (e.g: `["select * from " (Param "table-name")]`)
2. The `params.values/query->params-map` function takes a query and try to infer from the template-tags to return a param->value map. In our example above it should returns `{"table-name": "products"}`
3. And then, the `sql.params.substitute/substitute` will walk through the parsed-query from (1), and try to replace it with the value from (2)

Regarding snippet: currently in `params.values/query->params-map`, whenever we find a template-tag that need to be replaced as a snippet, we fetch the snippet from AppDB. So that `sql.params.substitute/substitute` could substitute with the snippet content.

### Solution
Treat snippet content like a native query:
- create `template_tags` column in `snippet` table
- In `params.values/query->params-map`, if a template-tag has target is a snippet, instead of returning the raw content like before, we'll parse the content of this snippet again with `params.parse/parse`([ref](https://github.com/metabase/metabase/blob/f44b3e993148585fdc424f8afbf6589484e7b0b5/src/metabase/driver/common/parameters/values.clj#L176)). 
  - Call `params.values/query->params-map` on the template-tags of snippet to get the param->value map.
  - With the parsed-query and param->value for snippet, in the later phase `sql.params.substitute/substitute`, we could substitute the param in snippet with its desired value by simply calling `sql.params.substitute/substitute` again. ([ref](https://github.com/metabase/metabase/blob/f44b3e993148585fdc424f8afbf6589484e7b0b5/src/metabase/driver/sql/parameters/substitute.clj#L27))

### How to test
1. New SQL query
2. Create a new snippet with content `rating > {{rating}}`
3. manually update the `template_tags` column of the added snippet
```json
{"rating":{"type":"number","name":"rating","id":"bf8e3f1f-cf69-e83b-d378-366c865ee221","display-name":"Rating","default":null}}
```
4. Back to the SQL console, create a query `select * from products where ` then insert the created snippet in.
5. Now an input for the `rating` variable should appear, gives it a value (e.g: 3), and try to run the query

### Note
We'll probably need to have a follow up PR for [this](https://github.com/metabase/metabase/pull/23691#discussion_r914694950)

P.s: FE changes are in #23526.

### CODE REVIEW

File: src/metabase/driver/common/parameters.clj
Comment (noahmoss) [inline src/metabase/driver/common/parameters.clj:48]: nit: minor copy editing
```suggestion
;; It is to be used as an intermediate state when substituting parameter in a Snippet.
```

File: src/metabase/driver/common/parameters.clj
Comment (noahmoss) [inline src/metabase/driver/common/parameters.clj:54]: nit: more minor copy editing
```suggestion
;; `param->value` is a map with with template tags parsed from the raw query as keys
```

File: test/metabase/models/native_query_snippet_test.clj
Comment (noahmoss) [inline test/metabase/models/native_query_snippet_test.clj:89]: nit: extra lines

**Review by noahmoss — APPROVED** [review]
👏 awesome, I think this looks great

File: src/metabase/util/schema.clj
Comment (metamben) [inline src/metabase/util/schema.clj:355]: I don't understand this comment. (I'm a newbie here.)

File: test/metabase/api/native_query_snippet_test.clj
Comment (metamben) [inline test/metabase/api/native_query_snippet_test.clj:245]:
```suggestion
                                                             :type         "dimension"
```

**Review by metamben — COMMENTED** [review]
I was expecting a test with a snippet having snippet template tags. As far as I understand the code, this recursion should work.

I like the code but it would be good to get a review from somebody with more experience in this area of code.

File: resources/migrations/000_migrations.yaml
Comment (qnkhuat) [reply resources/migrations/000_migrations.yaml:11951]: yes we can't, IIRC using `defaultValue` for text column will fail on MariaDB and MySQL. that's why we use this trick ([ref](https://github.com/metabase/metabase/blob/f44b3e993148585fdc424f8afbf6589484e7b0b5/resources/migrations/000_migrations.yaml#L2760))

File: test/metabase/api/native_query_snippet_test.clj
Comment (qnkhuat) [reply test/metabase/api/native_query_snippet_test.clj:245]: thanks!

File: src/metabase/util/schema.clj
Comment (qnkhuat) [reply src/metabase/util/schema.clj:355]: the schema defined here is primarily used for APIs, which it's basically is a looser version of the schema we use in MBQL.

The comment is a bit confusing. I'll update it

**Comment by qnkhuat** [comment]
> I was expecting a test with a snippet having snippet template tags

I didn't add it at first because we probably not going to support it on UI in the near future.
But I added one bc why not :) https://github.com/metabase/metabase/pull/23658/commits/1b6d229dfc8e7e00fc20218f94f29c06084aa10d

**Review by metamben — APPROVED** [review]
LGTM

**Comment by metamben** [comment]
FYI, I've got a PR for the followup on #23691, I'm just waiting for this to get merged.

**Review by bshepherdson — APPROVED** [review]

**Comment by qnkhuat** [comment]
Hey @metamben, this got merged, I think you have a follow-up for #23691?

**Pull request merged**
